source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,158 @@
+                 transformations) {}
+
+void FuzzerPassWrapVectorSynonym::Apply() {
+  ForEachInstructionWithInstructionDescriptor(
+      [this](opt::Function* /*unused*/, opt::BasicBlock* /*unused*/,
+             opt::BasicBlock::iterator instruction_iterator,
+             const protobufs::InstructionDescriptor& instruction_descriptor)
+          -> void {
+        // Only run fuzzer pass on supported scalar operation type instruction.
+        if (!TransformationWrapVectorSynonym::OpcodeIsSupported(
+                instruction_iterator->opcode()))
+          return;
+
+        assert(instruction_iterator->opcode() ==
+                   instruction_descriptor.target_instruction_opcode() &&
+               "The opcode of the instruction we might insert before must be "
+               "the same as the opcode in the descriptor for the instruction");
+
+        // Randomly decide whether to wrap it to a vector operation.
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfWrappingVectorSynonym())) {
+          return;
+        }
+
+        // It must be valid to insert an OpCompositeConstruct instruction
+        // before |instruction_iterator|.
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+                SpvOpCompositeConstruct, instruction_iterator)) {
+          return;
+        }
+        // Get the scalar type represented by the targeted instruction id.
+        uint32_t operand_type_id = instruction_iterator->type_id();
+
+        // Get a random vector size from 2 to 4.
+        uint32_t vector_size = GetFuzzerContext()->GetWidthOfWrappingVector();
+
+        // Randomly choose a position that target ids should be placed at.
+        // The position is in range [0, n - 1], where n is the size of the
+        // vector.
+        uint32_t position =
+            GetFuzzerContext()->GetRandomIndexForWrappingVector(vector_size);
+
+        // Target ids are the two scalar ids from the original instruction.
+        uint32_t target_id1 = instruction_iterator->GetSingleWordInOperand(0);
+        uint32_t target_id2 = instruction_iterator->GetSingleWordInOperand(1);
+
+        // Stores the ids of scalar constants.
+        std::vector<uint32_t> vec1_components;
+        std::vector<uint32_t> vec2_components;
+
+        // Width is specified in the first index for either OpTypeInt or
+        // OpTypeFloat.
+        uint32_t width = GetIRContext()
+                             ->get_def_use_mgr()
+                             ->GetDef(operand_type_id)
+                             ->GetSingleWordOperand(0);
+        // Get the scalar type.
+        auto type = GetIRContext()->get_type_mgr()->GetType(operand_type_id);
+        // Whether the constant is signed, not used for float type so default
+        // set to true.
+        bool is_signed_constant =
+            type->AsInteger() ? type->AsInteger()->IsSigned() : true;
```

Comment:
I believe this can go. See below for why.

Reply:
Done.